amdsharktuner/amdsharktuner/constraint_generator.py

```diff
@@ -7,14 +7,136 @@
 import z3  # type: ignore
 import math
 from abc import ABC, abstractmethod
-from typing import Iterator, Optional
+from typing import Iterator, Optional, Protocol, cast
+from dataclasses import dataclass, fields
 
 from iree.compiler import ir  # type: ignore
 from iree.compiler.dialects import iree_codegen, iree_gpu, linalg  # type: ignore
 
 from . import common, dispatch_constraints, dispatch_parser
 
 
+class Z3Vals(Protocol):
+    """Marker base class for values extracted from Z3 model."""
+
+
+class Z3Vars(Protocol):
+    """Marker base class for Z3 variables."""
```

Comment:
What do you mean by variables? Do you mean constants? See https://theory.stanford.edu/~nikolaj/programmingz3.html#sec-logical-interface